### ISSUE TITLE
Upsert table backfill enhancement: support externally partitioned data

### ISSUE THREAD

**Issue by rohityadav1993** [apache/pinot#12987]
## Problem
#6567 allows uploading a batch generated segment to Pinot upsert realtime table. Partitioned data is handled by defining the partition column in `segmentPartitionConfig.columnPartitionMap`. The addSegment flow uses the config to identify the partition value of the column in metadata.properties of the uplaoded segment and then assign the segment to instance based on the partition id and instance assignment zk metadata. 
This puts a restriction to define a partition column that is part of the table column(primary key) and also use only the configured [partition function](https://github.com/apache/pinot/blob/a5c728f549fe1be5560a88080caaa2063def3d87/pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/PartitionFunctionFactory.java#L31) for partitioning data during segment creating in batch job using `SegmentIndexCreationDriverImpl`

This restriction is not applicable for stream generated segments during realtime ingestion. The partition id is identified using the LLC segment name convention. The stream is partitioned externally and Pinot table does not need to be aware of the partitioned column/function.

## Proposal
Proposal to enhance `SegmentAssignment.assignSegment()` flow to rely on externally provided partition id for an uplaoded segment.

1. Need to persist the partition id as part of segment zk metadata.
2. Modify the `StrictRealtimeSegmentAssignment.assignSegment` to get partition id from zk metadata.
3. Provide partition id externally:
    1. Option 1: Provide partition id as http headers during segment upload
    2. Option 2: Provide partition id as part of uploaded segment metadata(not as columnPartitionMap) (metadata.properties)

Related: #10896, #11914

**Comment by Jackie-Jiang** [comment]
For real-time ingested data, the partition must match the upstream partition id to ensure the upsert assumption of all data of the same partition served by the same server, and I don't think we can loose this requirement.

`Partition function` is required for partition pruning. If partition pruning is not required, then we may allow custom partition id without a partition function.

**Comment by rohityadav1993** [comment]
> For real-time ingested data, the partition must match the upstream partition id to ensure the upsert assumption of all data of the same partition served by the same server, and I don't think we can loose this requirement.

Yes, this is not supposed to be changed, but here we are getting the partition id from upstream instead of Pinot applying any partition function on primary key column to find the partition id.
Similarly, we should have capability to provide a partition id for uploaded segment. Since the partitioning of data is being done externally and can provide the partition id for uplaoded segments as well.

> Partition function is required for partition pruning. If partition pruning is not required, then we may allow custom partition id without a partition function.

This should be doable with upstream configuration that one should use a partition function from Pinot's list and on a table's column.

Since Pinot has not restricted which partition function to use in the upstream for creating upsert tables, we have use cases that can use custom partition functions, these use-cases also don't need partition pruning.

**Comment by tibrewalpratik17** [comment]
> Provide partition id externally:
Option 1: Provide partition id as http headers during segment upload
Option 2: Provide partition id as part of uploaded segment metadata(not as columnPartitionMap) (metadata.properties)

IMO if we go for option-2, then we should be consistent to add this / update this metadata for all present segments too. Option-1 is better in that aspect as we already pass a lot of info as headers during segment upload and use each header as more of a config.

**Comment by rohityadav1993** [comment]
Another approach I believe can be utlized is defining a naming convention for uploaded segment similar to LLC. The segment name can capture the partition id. We already have a segment type as UPLOADED and `SegmentPartitionMetadataManager#getPartitionId` can be enhanced to extract partition id from name. 
This would not require any changes to existing contracts or zk metadata.
